File: spiral-matrix/spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int row=matrix.size(),col=matrix[0].size();
        int up=0,right=col-1,left=0,down=row-1;
        vector<int>ans;
        while(ans.size()<row*col){
            for(int col=left;col<=right;col++){
                ans.push_back(matrix[up][col]);
            }
            for(int row=up+1;row<=down;row++){
                ans.push_back(matrix[row][right]);
            }
            if(up!=down){
                for(int col=right-1;col>=left;col--){
                    ans.push_back(matrix[down][col]);
                }
            }
            if(left!=right){
                for(int row=down-1;row>up;row--){
                    ans.push_back(matrix[row][left]);
                }
            }
            left++;
            right--;
            up++;
            down--;
        }
        return ans;
    }
};

File: 54-spiral-matrix/spiral-matrix.py
class Solution:
    def spiralOrder(self, mat: List[List[int]]) -> List[int]:
        n = len(mat)
        m = len(mat[0])
        left = 0
        right = m -1
        top = 0
        bottom = n - 1
        ans = []
        while left<=right and top<=bottom:
            for j in range(left,right+1):
                ans.append(mat[top][j])
            top+=1

            for i in range(top,bottom+1):
                ans.append(mat[i][right])
            right-=1

            if top<=bottom:
                for j in range(right,left-1,-1):
                    ans.append(mat[bottom][j])
                bottom -= 1

            if left<=right:
                for i in range(bottom,top-1,-1):
                    ans.append(mat[i][left])
                left+=1
        return ans